classes/plugins/PKPPubIdPlugin.inc.php

```diff
@@ -0,0 +1,410 @@
+		$excluded = $pubObject->getData($excludeFormFieldName);
+		return $excluded;
+	}
+
+	/**
+	 * Get the error message for not unique pub id
+	 * @return string
+	 */
+	abstract function getNotUniqueErrorMsg();
+
+	/**
+	 * Verify form data.
+	 * @param $fieldName string The form field to be checked.
+	 * @param $fieldValue string The value of the form field.
+	 * @param $pubObject object
+	 * @param $contextId integer
+	 * @param $errorMsg string Return validation error messages here.
+	 * @return boolean
+	 */
+	function verifyData($fieldName, $fieldValue, &$pubObject, $contextId, &$errorMsg) {
+		// Verify pub id uniqueness.
+		if ($fieldName == $this->getSuffixFieldName()) {
+			if (empty($fieldValue)) return true;
+
+			// Construct the potential new pub id with the posted suffix.
+			$pubIdPrefix = $this->getSetting($contextId, $this->getSuffixFieldName());
+			if (empty($pubIdPrefix)) return true;
+			$newPubId = $pubIdPrefix . '/' . $fieldValue;
+
+			if($this->checkDuplicate($newPubId, $pubObject, $contextId)) {
+				return true;
+			} else {
+				$errorMsg = $this->getNotUniqueErrorMsg();
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Check whether the given pubId is valid.
+	 * @param $pubId string
+	 * @return boolean
+	 */
+	function validatePubId($pubId) {
+		return true; // Assume a valid ID by default;
+	}
+
+	/**
+	 * Return an array of publication object types and
+	 * the corresponding DAOs.
+	 * @return array
+	 */
+	abstract function getDAOs();
+
+	/**
+	 * Add the suffix element and the public identifier
+	 * to the object.
+	 * @param $hookName string
+	 * @param $params array ()
+	 */
+	function getAdditionalFieldNames($hookName, $params) {
+		$fields =& $params[1];
+		$formFieldNames = $this->getFormFieldNames();
+		foreach ($formFieldNames as $formFieldName) {
+			$fields[] = $formFieldName;
+		}
+		$daoFieldNames = $this->getDAOFieldNames();
+		foreach ($daoFieldNames as $daoFieldName) {
+			$fields[] = $daoFieldName;
+		}
+		return false;
+	}
+
+	/**
+	 * Return the object type.
+	 * @param $pubObject object
+	 * @return array
+	 */
+	function getPubObjectType($pubObject) {
+		$allowedTypes = $this->getPubObjectTypes();
+		$pubObjectType = null;
+		foreach ($allowedTypes as $allowedType) {
+			if (is_a($pubObject, $allowedType)) {
+				$pubObjectType = $allowedType;
+				break;
+			}
+		}
+		if (is_null($pubObjectType)) {
+			// This must be a dev error, so bail with an assertion.
+			assert(false);
+			return null;
+		}
+		return $pubObjectType;
+	}
+
+	/**
+	 * Set and store a public identifier.
+	 * @param $pubObject object
+	 * @param $pubObjectType string As returned from self::getPubObjectType()
+	 * @param $pubId string
+	 * @return string
+	 */
+	function setStoredPubId(&$pubObject, $pubObjectType, $pubId) {
+		$daos = $this->getDAOs();
+		$dao = DAORegistry::getDAO($daos[$pubObjectType]);
+		$dao->changePubId($pubObject->getId(), $this->getPubIdType(), $pubId);
```

Comment:
Do we want to use an interface or subclass of DAOs that defines some of these methods that are assumed to exist? Might be nice.